src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -70,7 +69,37 @@ internal virtual void GetPolicies(JsonSerializerOptions options)
         {
             if (RuntimePropertyType.IsArray)
             {
-                EnumerableConverter = s_jsonEnumerableConverter;
+                EnumerableConverter = new DefaultArrayConverter();
+            }
+            else if (typeof(IEnumerable).IsAssignableFrom(RuntimePropertyType))
+            {
+                Type elementType = JsonClassInfo.GetElementType(RuntimePropertyType);
+
+                if (elementType == null)
+                {
+                    return;
+                }
+
+                if (RuntimePropertyType == typeof(IEnumerable<>).MakeGenericType(elementType))
+                {
+                    EnumerableConverter = new DefaultIEnumerableTConverter();
+                }
+                else if (RuntimePropertyType == typeof(IList<>).MakeGenericType(elementType))
+                {
+                    EnumerableConverter = new DefaultIListTConverter();
+                }
+                else if (RuntimePropertyType == typeof(ICollection<>).MakeGenericType(elementType))
+                {
+                    EnumerableConverter = new DefaultICollectionTConverter();
+                }
+                else if (RuntimePropertyType == typeof(IReadOnlyCollection<>).MakeGenericType(elementType))
+                {
+                    EnumerableConverter = new DefaultIReadOnlyCollectionTConverter();
+                }
+                else if (RuntimePropertyType == typeof(IReadOnlyList<>).MakeGenericType(elementType))
+                {
+                    EnumerableConverter = new DefaultIReadOnlyListTConverter();
```

Comment:
Do we really need this many specialized types? There's an overhead associated with every new type that gets loaded into the system (working set, etc.)

I understand that we want to prevent someone to get an `IEnumerable<T>` and cast it to `IList<T>` because they figured out the implementation detail, but in the past we decided we didn't care about breaking that kind of misuse (can't find the issue #, but I for sure remember some).

Reply:
> in the past we decided we didn't care about breaking that kind of misuse

In what area?

There are two types of leaking:
1) minor leak: returning internal type with extra public interface(s)
2) major leak: returning a public class, like `List<T>`

For either it may be considered a breaking change if we want to change our implementation later.

Reply:
The case where we return a different type (that may have different interfaces).

E.g. we changed the implementation of `public static IEnumerable<TResult> Empty<TResult>() => Array.Empty<TResult>();` to `public static IEnumerable<TResult> Empty<TResult>() => EmptyPartition<TResult>.Instance;` in LINQ and nobody is going to shed a tear for customers who did hard casts of `Enumerable.Empty<T>()` to `T[]`.

We wouldn't be able to do that in .NET Framework but this specific case feels exactly why we have .NET Core - we don't need to do unnatural things just to avoid breaking badly behaving apps in the future.

Reply:
So, what's the conclusion here? Always new up the `DefaultIListTConverter` rather than specialized types and use the concrete `List<T>`?

Reply:
> So, what's the conclusion here? 

See https://github.com/dotnet/corefx/pull/36756#discussion_r275016264.